README.md
![TheAudioDBSync](icon.png)

TheAudioDBSync addon allows you to retrieve Album and track ratings from [theaudiodb.com](http://www.theaudiodb.com) and update your Kodi Library with the new values.

For more information you can have a look at the wiki entry:

[Add-on:TheAudioDBSync](https://github.com/robwebset/script.theaudiodb.sync/wiki)

You can install this addon from: [Repository:URepo](http://www.urepo.org/)

To ask questions or comment about TheAudioDBSync, please use the forum at [URepo.org](http://www.urepo.org/forum/viewtopic.php?t=1772)

__Donations__

If you have found TheAudioDBSync useful and would like to help support future development and bug fixes, then please feel free to make a small donation.

[![Donation](https://www.paypalobjects.com/en_GB/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=Q9KP578EJJ7L6)


changelog.txt
v1.2.0 (2018-06-01)
- Tidy Up

v1.1.0 (2017-12-31)
- Set default user when using pre-configured users

v1.0.9 (2017-12-31)
- Do not require username to be set when using pre-configured users

v1.0.8 (2017-10-29)
- Allow download of ratings from pre-configured users
- Add Hebrew Language

v1.0.6 (2017-10-22)
- Fix script error on saving summary log

v1.0.5 (2017-04-27)
- Add option to log a summary of the updates performed

v1.0.4 (2017-04-17)
- Fix upload support character encoding issue

v1.0.3 (2017-04-06)
- Support Ratings upload

v1.0.2 (2017-01-23)
- Prevent script error if theaudiodb.com is unreachable

v1.0.1 (2017-01-16)
- Add option to schedule synchronisation

v1.0.0 (2017-01-10)
- Initial Version


default.py
# -*- coding: utf-8 -*-
import time
import xbmcgui
import xbmcaddon

# Import the common settings
from resources.lib.settings import log
from resources.lib.settings import Settings
from resources.lib.sync import LibrarySync
from resources.lib.summary import Summary


ADDON = xbmcaddon.Addon(id='script.theaudiodb.sync')


##################################
# Main of TheAudioDBSync Script
##################################
if __name__ == '__main__':
    log("TheAudioDBSync Starting %s" % ADDON.getAddonInfo('version'))

    # Get the username
    username = Settings.getUsername()

    if username in [None, ""]:
        # Show a dialog detailing that the username is not set
        xbmcgui.Dialog().ok(ADDON.getLocalizedString(32001), ADDON.getLocalizedString(32005))
    else:
        dialogSummary = ''

        # Make the call to upload any ratings that have changed
        numAlbumRatingsUploaded = -1
        if Settings.isUploadAlbumRatings():
            numAlbumRatingsUploaded = LibrarySync.checkForChangedAlbumRatings(username, True)

        numTrackRatingsUploaded = -1
        if Settings.isUploadTrackRatings():
            numTrackRatingsUploaded = LibrarySync.checkForChangedTrackRatings(username, True)

        if (numTrackRatingsUploaded != -1) or (numAlbumRatingsUploaded != -1):
            if numAlbumRatingsUploaded < 0:
                numAlbumRatingsUploaded = 0
            if numTrackRatingsUploaded < 0:
                numTrackRatingsUploaded = 0

            # Display a summary of what was performed
            dialogSummary = "%d %s\n%d %s\n" % (numAlbumRatingsUploaded, ADDON.getLocalizedString(32030), numTrackRatingsUploaded, ADDON.getLocalizedString(32031))

        # Only check for resync if it is enabled
        if Settings.isUpdateAlbumRatings() or Settings.isUpdateTrackRatings():
            performResync = True
            # Before performing the resync, check when the last time a resync was done, we want
            # to try and discourage people doing too many resyncs in quick succession
            lastResyncTimeStr = Settings.getLastSyncTime()
            if lastResyncTimeStr not in [None, ""]:
                currentTime = int(time.time())
                # check if the last resync was within 5 minutes
                if currentTime < (int(lastResyncTimeStr) + 300):
                    performResync = xbmcgui.Dialog().yesno(ADDON.getLocalizedString(32001), ADDON.getLocalizedString(32016))

            if performResync:
                # Perform the resync operation and display the status
                numAlbumsUpdated, numTracksUpdated = LibrarySync.syncToLibrary(username, True)

                # Display a summary of what was performed
                dialogSummary += "%d %s\n%d %s" % (numAlbumsUpdated, ADDON.getLocalizedString(32010), numTracksUpdated, ADDON.getLocalizedString(32011))

        # Check if the summary should be saved
        if Settings.isSummaryLogEnabled():
            summary = Summary()
            summary.saveToDisk()
            del summary

        if dialogSummary not in [None, ""]:
            xbmcgui.Dialog().ok(ADDON.getLocalizedString(32001), dialogSummary)

    log("TheAudioDBSync Finished")


service.py
# -*- coding: utf-8 -*-
import time
import xbmcaddon

# Import the common settings
from resources.lib.settings import log
from resources.lib.settings import Settings
from resources.lib.sync import LibrarySync
from resources.lib.summary import Summary


ADDON = xbmcaddon.Addon(id='script.theaudiodb.sync')


##################################
# Main of TheAudioDBSync Script
##################################
if __name__ == '__main__':
    log("TheAudioDBSync Service Starting %s" % ADDON.getAddonInfo('version'))

    # Get the username
    username = Settings.getUsername()

    # If the username is not set, then nothing to do yet
    if username not in [None, ""]:
        performResync = False

        if Settings.isUploadRatingsOnStartup():
            LibrarySync.checkForChangedTrackRatings(username, False)
            LibrarySync.checkForChangedAlbumRatings(username, False)

        # Only check for resync if it is enabled
        if Settings.isUpdateAlbumRatings() or Settings.isUpdateTrackRatings():
            nextResyncTime = Settings.getNextScheduledResyncTime()

            if nextResyncTime not in [None, "", "0"]:
                log("Service: Next Sync time is %d" % nextResyncTime)
                currentTime = int(time.time())
                log("Service: Current time is %d" % currentTime)
                # check if the last resync was within 5 minuted
                if currentTime > nextResyncTime:
                    performResync = True

        if performResync:
            # Check if the status should be displayed
            displayProgress = Settings.isScheduleDisplayProgress()
            # Perform the resync operation and display the status
            numAlbumsUpdated, numTracksUpdated = LibrarySync.syncToLibrary(username, displayProgress)

        # Check if the summary should be saved
        if Settings.isSummaryLogEnabled():
            summary = Summary()
            summary.saveToDisk()
            del summary

    log("TheAudioDBSync Service Finished")


clearhistory.py
# -*- coding: utf-8 -*-
import traceback
import xbmc
import xbmcaddon
import xbmcvfs
import xbmcgui

# Import the common settings
from resources.lib.settings import log

ADDON = xbmcaddon.Addon(id='script.theaudiodb.sync')
ADDON_ID = ADDON.getAddonInfo('id')


#########################
# Main
#########################
if __name__ == '__main__':
    log("AudioDBSync: Clear History Called (version %s)" % ADDON.getAddonInfo('version'))

    trackRatingsPath = xbmc.translatePath('special://profile/addon_data/%s/trackRatings.json' % ADDON_ID).decode("utf-8")
    albumRatingsPath = xbmc.translatePath('special://profile/addon_data/%s/albumRatings.json' % ADDON_ID).decode("utf-8")

    if xbmcvfs.exists(trackRatingsPath):
        try:
            log("AudioDBSync: Removing file %s" % trackRatingsPath)

            xbmcvfs.delete(trackRatingsPath)
        except:
            log("AudioDBSync: %s" % traceback.format_exc(), xbmc.LOGERROR)

    if xbmcvfs.exists(albumRatingsPath):
        try:
            log("AudioDBSync: Removing file %s" % albumRatingsPath)

            xbmcvfs.delete(albumRatingsPath)
        except:
            log("AudioDBSync: %s" % traceback.format_exc(), xbmc.LOGERROR)

    xbmcgui.Dialog().ok(ADDON.getLocalizedString(32001), ADDON.getLocalizedString(32034))


resources/lib/summary.py
# -*- coding: utf-8 -*-
import traceback
import xbmc
import xbmcaddon
import xbmcvfs


ADDON = xbmcaddon.Addon(id='script.theaudiodb.sync')
ADDON_ID = ADDON.getAddonInfo('id')

from settings import log


##############################
# Stores Progress Summary
##############################
class Summary():
    # Store the items globally, all instances should store their items together until they are saved to disk
    _items = []
    F_UNKNOWN = 'UNKNOWN'
    F_UPLOAD = 'UPLOAD'
    F_DOWNLOAD = 'DOWNLOAD'
    A_UNKNOWN = 'UNKNOWN'
    A_TRACK = 'TRACK'
    A_ALBUM = 'ALBUM'

    def __init__(self):
        self.current = None
        self.clearCurrent()

    def saveCurrent(self):
        Summary._items.append(self.current.copy())
        self.clearCurrent()

    def clearCurrent(self):
        self.current = {'function': Summary.F_UNKNOWN, 'area': Summary.A_UNKNOWN, 'artist': '', 'title': '', 'oldRating': -1, 'newRating': -1, 'result': 'No change'}

    def saveToDisk(self):
        log("saveToDisk: Saving summary to disk")
        logPath = xbmc.translatePath('special://profile/addon_data/%s/sync_summary.log' % ADDON_ID).decode("utf-8")

        if len(Summary._items) < 1:
            log("saveToDisk: No items in the summary")
            return

        allData = 'OPERATION|AREA|ARTIST|TITLE|OLD or TOTAL RATING|NEW RATING|RESULT\n'
        for item in Summary._items:
            allItems = []
            if item['function'] is None:
                allItems.append('')
            else:
                allItems.append(item['function'])
            if item['area'] is None:
                allItems.append('')
            else:
                allItems.append(item['area'])
            if item['artist'] is None:
                allItems.append('')
            else:
                allItems.append(item['artist'])
            if item['title'] is None:
                allItems.append('')
            else:
                allItems.append(item['title'])
            if item['oldRating'] is None:
                allItems.append('')
            else:
                allItems.append(str(item['oldRating']))
            if item['newRating'] is None:
                allItems.append('')
            else:
                allItems.append(str(item['newRating']))
            if item['result'] is None:
                allItems.append('')
            else:
                allItems.append(item['result'])
            allData += '|'.join(allItems)
            allData += '\n'

        try:
            logFile = xbmcvfs.File(logPath, 'wb')
            logFile.write(bytearray(allData, 'utf_8'))
            logFile.close()
        except:
            log("saveToDisk: Failed to write file: %s" % logPath, xbmc.LOGERROR)
            log("saveToDisk: %s" % traceback.format_exc(), xbmc.LOGERROR)

        Summary._items = []

    def reset(self):
        Summary._items = []


resources/lib/settings.py
# -*- coding: utf-8 -*-
import time
import xbmc
import xbmcaddon

ADDON = xbmcaddon.Addon(id='script.theaudiodb.sync')
ADDON_ID = ADDON.getAddonInfo('id')


# Common logging module
def log(txt, loglevel=xbmc.LOGDEBUG):
    if (ADDON.getSetting("logEnabled") == "true") or (loglevel != xbmc.LOGDEBUG):
        if isinstance(txt, str):
            txt = txt.decode("utf-8")
        message = u'%s: %s' % (ADDON_ID, txt)
        xbmc.log(msg=message.encode("utf-8"), level=loglevel)


##############################
# Stores Various Settings
##############################
class Settings():
    @staticmethod
    def getKodiVersion():
        kodiVer = xbmc.getInfoLabel('system.buildversion')
        majorVersion = 17
        minorVersion = 0
        try:
            majorVersion = int(kodiVer.split(".", 1)[0])
        except:
            log("Failed to get major version, using %d" % majorVersion)
        try:
            minorSplit = kodiVer.split(".", 2)[1]
            # Non GA versions there are bits after a minus
            minorVersion = int(minorSplit.split("-", 1)[0])
        except:
            log("Failed to get minor version, using %d" % minorVersion)

        return majorVersion, minorVersion

    @staticmethod
    def isSummaryLogEnabled():
        return ADDON.getSetting("summaryLogEnabled") == "true"

    @staticmethod
    def getUsername():
        username = ADDON.getSetting("username")
        # Check if a default download user is set
        if username in [None, '']:
            username = Settings.getSpecialistDownloadUser()
        return username

    @staticmethod
    def getApiToken():
        return ADDON.getSetting("apiToken")

    @staticmethod
    def isUseArtistDetails():
        return ADDON.getSetting("useArtistDetails") == "true"

    @staticmethod
    def isUpdateAlbumRatings():
        return ADDON.getSetting("updateAlbumRatings") == "true"

    @staticmethod
    def isUpdateTrackRatings():
        return ADDON.getSetting("updateTrackRatings") == "true"

    @staticmethod
    def isUploadAlbumRatings():
        if Settings.getApiToken() in [None, ""]:
            return False
        return ADDON.getSetting("uploadAlbumRatings") == "true"

    @staticmethod
    def isUploadTrackRatings():
        if Settings.getApiToken() in [None, ""]:
            return False
        return ADDON.getSetting("uploadTrackRatings") == "true"

    @staticmethod
    def isUploadRatingsOnStartup():
        if Settings.getApiToken() in [None, ""]:
            return False
        return ADDON.getSetting("uploadRatingsOnStartup") == "true"

    @staticmethod
    def doNotUploadZeroRatings():
        if Settings.getApiToken() in [None, ""]:
            return True
        return ADDON.getSetting("doNotUploadZeroRatings") == "true"

    @staticmethod
    def getLastSyncTime():
        return ADDON.getSetting("lastSyncTime")

    @staticmethod
    def setLastSyncTime(setDisplayTime=False):
        # Get the current EPOCH time
        epoch = str(int(time.time()))
        ADDON.setSetting("lastSyncTime", epoch)
        # Also set the display time if required
        if setDisplayTime:
            ADDON.setSetting("lastSyncDisplay", time.strftime('%H:%M:%S %Y-%m-%d', time.localtime(float(epoch))))

    @staticmethod
    def getNextScheduledResyncTime():
        index = int(ADDON.getSetting("scheduleInterval"))
        if index == 0:
            return None

        lastResyncStr = Settings.getLastSyncTime()
        if lastResyncStr in [None, "", "0"]:
            log("getNextScheduledResyncTime: No sync time set, using current")
            # Set the current time as the last resync as this will trigger in the future
            Settings.setLastSyncTime(False)
            return None

        if index == 1:
            # Weekly (604,800 seconds)
            return int(lastResyncStr) + 604800
        elif index == 2:
            # Fortnightly (1,209,600 seconds)
            return int(lastResyncStr) + 1209600
        elif index == 3:
            # Monthly (2,592,000 seconds) so 30 days
            return int(lastResyncStr) + 2592000

    @staticmethod
    def isScheduleDisplayProgress():
        return ADDON.getSetting("scheduleDisplayProgress") == "true"

    @staticmethod
    def getSpecialistDownloadUser():
        downloadUser = None
        if ADDON.getSetting("updateFromSpecialistUser") == "true":
            downloadUser = ADDON.getSetting("specialistUser")
        return downloadUser


resources/lib/library.py
# -*- coding: utf-8 -*-
import sys
import xbmc

if sys.version_info >= (2, 7):
    import json
else:
    import simplejson as json

# Import the common settings
from settings import log
from settings import Settings


# Class to handle calls to the Kodi Music Library
class MusicLibrary():
    def __init__(self):
        # Each version of Kodi stores the ratings slightly differently so before
        # we do anything find out which version is running
        self.kodiMajorVersion, kodiMinorVersion = Settings.getKodiVersion()

        # Looks like Jarvis (v16.0) and earlier are 1-5, otherwise 1-10
        self.ratingDivisor = 1
        if (self.kodiMajorVersion < 16) or ((self.kodiMajorVersion < 17) and (kodiMinorVersion < 1)):
            log("Detected version earlier than 16.1, using rating range 1-5")
            self.ratingDivisor = 2

        self.additionalTrackValues = ''
        self.ratingName = 'rating'
        if self.kodiMajorVersion >= 17:
            self.additionalTrackValues = ', "userrating"'
            self.ratingName = 'userrating'
        # Check if we may want to use the Artist Details as well
        if Settings.isUseArtistDetails():
            self.additionalTrackValues = '%s, "title", "artist"' % self.additionalTrackValues

    # Get details about every track in the Kodi Music Library
    def getLibraryTracks(self):
        json_query = xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "AudioLibrary.GetSongs", "params": {"properties": ["albumartist", "musicbrainztrackid", "rating"%s]  }, "id": "libSongs"}' % self.additionalTrackValues)
        json_response = json.loads(json_query)

        libraryTracks = []
        if ("result" in json_response) and ('songs' in json_response['result']):
            libraryTracks = json_response['result']['songs']
            # Check for the case where there is no artist, but there is an album artist
            for track in libraryTracks:
                if ('artist' not in track) or (len(track['artist']) < 1) or (len(''.join(track['artist'])) < 1):
                    if ('albumartist' in track) and (len(track['albumartist']) > 0):
                        track['artist'] = track['albumartist']
                # We don't want the album artist left in the data
                if ('albumartist' in track):
                    del track['albumartist']
                # If we are running on a version earlier than v17, populate the rating
                if (self.kodiMajorVersion < 17) and ('userrating' not in track) and ('rating' in track):
                    track['userrating'] = int(track['rating'] * self.ratingDivisor)

        log("MusicLibrary: Retrieved a total of %d tracks" % len(libraryTracks))
        return libraryTracks

    # Get details about every track in the Kodi Music Library
    def getLibraryAlbums(self):
        json_query = xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "AudioLibrary.GetAlbums", "params": {"properties": ["musicbrainzalbumid", "rating"%s]  }, "id": "libSongs"}' % self.additionalTrackValues)
        json_response = json.loads(json_query)

        libraryAlbums = []
        if ("result" in json_response) and ('albums' in json_response['result']):
            libraryAlbums = json_response['result']['albums']

            for album in libraryAlbums:
                # If we are running on a version earlier than v17, populate the rating
                if (self.kodiMajorVersion < 17) and ('userrating' not in album) and ('rating' in album):
                    album['userrating'] = int(album['rating'] * self.ratingDivisor)

        log("MusicLibrary: Retrieved a total of %d albums" % len(libraryAlbums))
        return libraryAlbums

    # Update the rating for a given track
    def updateLibraryTrackRatings(self, libTrack, rating, totalRating):
        # Make sure the song Id is set, otherwise we have nothing to update
        if (libTrack in [None, ""]) or ('songid' not in libTrack) or (libTrack['songid'] in [None, ""]):
            return

        if (rating in [None, ""]) and (totalRating in [None, ""]):
            return

        songid = libTrack['songid']

        valuesToSet = ""
        existingRating = 0
        if 'rating' in libTrack:
            existingRating = int(libTrack['rating'])

        if self.kodiMajorVersion >= 17:
            # Check if the main rating needs to be updated
            if (totalRating not in [None, "", 0]) and (totalRating != existingRating):
                valuesToSet = ', "rating": %d' % totalRating
            # Check if the user rating needs updating
            existingUserRating = 0
            if 'userrating' in libTrack:
                existingUserRating = int(libTrack['userrating'])
            if (rating not in [None, "", 0]) and (rating != existingUserRating):
                valuesToSet = '%s, "userrating": %d' % (valuesToSet, rating)
        else:
            # For older kodi versions we update the rating with the user rating
            if rating not in [None, "", 0]:
                correctedRating = int(rating / self.ratingDivisor)
                if correctedRating != existingRating:
                    valuesToSet = ', "rating": %d' % correctedRating

        trackUpdated = False
        # Check if we have any values to update
        if valuesToSet in [None, ""]:
            log("updateLibraryTrackRatings: no ratings to update, songid:%s, rating:%s, totalRating:%s" % (songid, str(rating), str(totalRating)))
        else:
            log("updateLibraryTrackRatings: updating songid %s with%s" % (songid, valuesToSet))
            setJson = '{"jsonrpc": "2.0", "method": "AudioLibrary.SetSongDetails", "params": { "songid": %s%s }, "id": "libSongs"}' % (songid, valuesToSet)
            xbmc.executeJSONRPC(setJson)
            trackUpdated = True

        return trackUpdated

    # Update the rating for a given album
    def updateLibraryAlbumRatings(self, libAlbum, rating, totalRating):
        # Make sure the song Id is set, otherwise we have nothing to update
        if (libAlbum in [None, ""]) or ('albumid' not in libAlbum) or (libAlbum['albumid'] in [None, ""]):
            return

        if (rating in [None, ""]) and (totalRating in [None, ""]):
            return

        albumid = libAlbum['albumid']

        valuesToSet = ""
        existingRating = 0
        if 'rating' in libAlbum:
            existingRating = int(libAlbum['rating'])

        if self.kodiMajorVersion >= 17:
            # Check if the main rating needs to be updated
            if (totalRating not in [None, "", 0]) and (totalRating != existingRating):
                valuesToSet = ', "rating": %d' % totalRating
            # Check if the user rating needs updating
            existingUserRating = 0
            if 'userrating' in libAlbum:
                existingUserRating = int(libAlbum['userrating'])
            if (rating not in [None, "", 0]) and (rating != existingUserRating):
                valuesToSet = '%s, "userrating": %d' % (valuesToSet, rating)
        else:
            # For older kodi versions we update the rating with the user rating
            if rating not in [None, "", 0]:
                correctedRating = int(rating / self.ratingDivisor)
                if correctedRating != existingRating:
                    valuesToSet = ', "rating": %d' % correctedRating

        albumUpdated = False
        # Check if we have any values to update
        if valuesToSet in [None, ""]:
            log("updateLibraryAlbumRatings: no ratings to update, albumid:%s, rating:%s, totalRating:%s" % (albumid, str(rating), str(totalRating)))
        else:
            log("updateLibraryAlbumRatings: updating albumid %s with%s" % (albumid, valuesToSet))
            setJson = '{"jsonrpc": "2.0", "method": "AudioLibrary.SetAlbumDetails", "params": { "albumid": %s%s }, "id": "libSongs"}' % (albumid, valuesToSet)
            xbmc.executeJSONRPC(setJson)
            albumUpdated = True

        return albumUpdated
